### ISSUE TITLE
[feature suggestion] Support component injection in abstract base class

### ISSUE THREAD

**Issue by incaseoftrouble** [mapstruct/mapstruct#2257]
Another one :-)

Suppose I have a mapper like this:
```java
@Mapper // with proper configuration of componentModel, injectionStrategy = CONSTRUCTOR, etc.
public abstract class Mapper {
  @Autowired
  private SpringComponent component;

  // mapping methods
}
```
This works, however using the recommended constructor injection

```java
@Mapper // with proper configuration of componentModel etc.
public abstract class Mapper {
  private final SpringComponent component;

  @Autowired
  Mapper(SpringComponent component) {
    this.component = component;
  }

  // mapping methods
}
```
does not work, since the generated mapper always a default constructor to be present. I suggest that if a single `@Autowired` constructor (or equivalent for other component models) is found, it takes the arguments and adds it to the generated constructor (including `@Qualifier` annotations etc.)

This is sort of low priority because using field injection on the base mapper still works, but would be nice to have I think.

**Comment by filiphr** [comment]
I think that we had this request when we were implementing this construction injection strategy. However, it is a bit tricky to get it right for all the different component models.

I think that the complexity of the implementation is not worth it.

And you can always use setter injection to make your mappers more easily testable.

e.g.

```
@Mapper // with proper configuration of componentModel etc.
public abstract class Mapper {
  private final SpringComponent component;
  
  @Autowire
  void setComponent(SpringComponent component) {
    this.component = component;
  }

  // mapping methods
}
```

If you are interested you can give it a go at doing a draft for this so we can see how it will look like.

By the way you don't need to add `@Autowired` on constructors anymore. Spring handles that when you have a single constructor.

**Comment by incaseoftrouble** [comment]
> By the way you don't need to add @Autowired on constructors anymore. Spring handles that when you have a single constructor.

I was aware of that, just wanted to emphasize ;-)

> However, it is a bit tricky to get it right for all the different component models.

Mh I guessed so. It should be rather easy for spring, right? You could literally just copy the constructor signature and append whatever is needed.

Admittedly, I don't know much about code generation / introspection and other component models. You probably would want to support that for all component models, right?

**Comment by filiphr** [comment]
Yes for Spring it would be rather easy as there is only a single constructor. However, `cdi` currently needs to have an empty constructor for `ApplicationScoped` beans. If we are doing something we should make it work for all.

Perhaps we should just accept and copy the constructor defined in the abstract base class. Not sure how easy it is with the current code that we have as the way it is done is a bit special :). What do you think @sjaakd, is it worth to add this? For cdi we can leave it up to the normal compilation to fail as the users will have to provide 2 constructors there as well, but that is OK I think

**Comment by AZch** [comment]
Hi, any updates?

**Comment by nbrugger-tgm** [comment]
Hi, I am advocating for this feature too.
Atm is is very unclear how to build (possibly multiple) nested Mappers for testing.

Lets assume there is this structure
```
Mapper A
|-MapperB
|  |-ServiceA
|-MapperD
```
Assuming there is a `@Mock` or `@Spy` for MapperB, ServiceA and MapperD it is quite hard to find a useable approach to test MapperA

Hope/Expectation:
```java
@Mock
ServiceA sa;
@Spy
MapperB mb = new MapperB(sa);
@Spy
MapperD mb = new MapperB();
@InjectMocks
MapperA ma;//or = new MapperA(mb,md);
```

While this holds still inconveniences (due to limitations from mockito) it would be a lot better than existing solutions:
* Using reflections to set the fields (big oucht + max 1 layer)
* Creating a setter for each field (who knows maybe there are some Mapper that need many services/mappers) because they would just be used for testing
* I dont know many other ways

**Comment by TheAndreyy** [comment]
Hey, I stumbled on this issue when having the same problem with Mapstruct. I've submitted PR to add this feature, waiting for your review :)

**Comment by PhilShpy** [comment]
A little upgrade of @filiphr 's solution:
with Lombok it's possible to use

```
@Mapper // with proper configuration of componentModel etc.
public abstract class Mapper {
  
  @Setter(onMethod_={@Autowired})
  private final SpringComponent component;

  // mapping methods
}
```

**Comment by fjakimowicz-da** [comment]
i dont get the issue here, even if is needed to keep consistency across different dependency injection frameworks. If a mapper is based on a superclass and is defined as componentModel = "spring", mapstruct could inherit the parent constructor without side effects i would guess, at least provide an extra annotation @MapperInheritConstructor o whatever so the users could specifically force mapstruct to do so. What would be the inconvenince on doing so?
